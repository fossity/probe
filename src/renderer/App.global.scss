/*
 * 3rd Party Libraries
 * @NOTE: Prepend a `~` to css file paths that are in your node_modules
 *        See https://github.com/webpack-contrib/sass-loader#imports
 */
@import '~react-virtualized/styles.css';
@import '~react-virtualized-tree/lib/main.css';
@import '~material-icons/css/material-icons.css';
// @import '~remixicon/fonts/remixicon.css';

/* fonts & vars */
@import './theme/fonts.scss';
@import './theme/variables.scss';

/* theme */
@import './theme/Main.global.scss';
@import './theme/Layout.global.scss';
@import './theme/SplitPane.global.scss';
@import './theme/Dialog.global.scss';

/* utils */
@import './theme/Utils.global.scss';

/* components */
@import './features/workspace/pages/Workspace/Workspace';
@import './features/workspace/pages/ProjectScan/ProjectScan';
@import './features/workspace/pages/ProjectSettings/ProjectSettings';
@import './features/workspace/pages/ProjectObfuscation/ProjectObfuscation';
@import './features/workspace/pages/ProjectSummary/ProjectSummary';
@import 'features/workspace/pages/ProjectResult/ProjectResult';
@import 'features/workspace/pages/Components/ProjectInfo/ProjectInfo';

/* ðŸ‘® css rules are forbidden here ðŸ‘® */
