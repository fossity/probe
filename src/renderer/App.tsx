import React from 'react';
import { createRoot } from 'react-dom/client';
import { HashRouter, Route, Routes } from 'react-router-dom';
import { Provider } from 'react-redux';
import i18next from 'i18next'
import { showTranslations } from 'translation-check'

import AppConfig from '@config/AppConfigModule';
import { DialogProvider } from '@context/DialogProvider';
import AppProvider from '@context/AppProvider';
import store from '@store/store';
import { createTheme, ThemeProvider, StyledEngineProvider, Theme } from '@mui/material/styles';

import { AppI18n } from '@shared/i18n';
import { IpcChannels } from '@api/ipc-channels';
import { userSettingService } from '@api/services/userSetting.service';
import WorkspaceModule from './features/workspace';
import AboutModule from './features/about';
import './App.global.scss';

declare module '@mui/styles/defaultTheme' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface DefaultTheme extends Theme {}
}

export default class App {
  /**
   * Initialize React application.
   *
   * @returns {JSX.Element}
   */
  public async setup(): Promise<void | Element | React.Component> {
    const { LNG } = await userSettingService.get();

    AppI18n.setLng(LNG);
    AppI18n.init();

    this.setTitle();
    const theme = this.loadTheme();

    const app = (
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <HashRouter>
              <Provider store={store}>
                <DialogProvider>
                  <AppProvider>
                    <React.Suspense fallback="">
                        <Routes>
                          <Route index element={<WorkspaceModule />} />
                          <Route path="/workspace/*" element={<WorkspaceModule />} />
                          <Route path="/about" element={<AboutModule />} />
                        </Routes>
                      </React.Suspense>
                  </AppProvider>
                </DialogProvider>
              </Provider>
            </HashRouter>
          </ThemeProvider>
        </StyledEngineProvider>
    );

    this.setupAppMenuListeners();

    const container = document.getElementById('root')!;
    const root = createRoot(container);
    root.render(app);
  }

  private async setTitle() {
    const appInfo = await window.app.getInfo()
    document.title = `${AppConfig.APP_NAME} (${appInfo.version})`;
  }

  setupAppMenuListeners() {
    window.electron.ipcRenderer.on(IpcChannels.MENU_OPEN_TRANSLATION_MANAGER, async (event) => {
     showTranslations(i18next);
    });
  }

  private loadTheme(): Theme {
    const primary = '#307FE2';
    const secondary = '#EBF3FD';

    const theme = createTheme({
      palette: {
        mode: 'light',
        primary: {
          main: primary,
        },
        secondary: {
          main: secondary,
          contrastText: '#307FE2',
        },
        success: {
          main: '#02BD63',
        },
      },
      typography: {
        button: {
          fontWeight: 600,
          textTransform: 'none',
        },
        fontFamily: '"Inter", "Roboto", "Helvetica", "Arial", sans-serif',
      },
      components: {
        MuiButton: {
          defaultProps: {
            disableElevation: true,
          },
          styleOverrides: {
            root: {
              borderRadius: 10,
            }
          }
        },
        MuiChip: {
          styleOverrides: {
            root: {
              borderRadius: 6,
            }
          }
        },
        MuiFilledInput: {
          styleOverrides: {
            root: {
              backgroundColor: 'white',
              borderRadius: 10,

              // TODO: we need to remove "!important" from rules. Currently, default style is more specific that custom styles. Could be controversial in the future
              '&.Mui-disabled': {

              },

              '&.Mui-error': {
                borderColor: '#d32f2f !important',
              },
              '&.Mui-focused.MuiInputBase-adornedStart > .MuiSvgIcon-root:first-child': {
                color: primary,
              },
            },
          },
        },
        MuiTableHead: {
          styleOverrides: {
            root: {
              '& .MuiTableCell-root': {
                color: '#27272A',
                backgroundColor: '#FAFAFA !important',
                fontWeight: 700,
              },
            },
          },
        },
      },
    });

    theme.shadows[1] = '0px 1px 3px 0px #0000001A; 1px 0px 2px 0px #0000000F';
    return theme;
  }
}
