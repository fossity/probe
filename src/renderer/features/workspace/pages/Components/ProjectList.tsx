/* eslint-disable react/destructuring-assignment */
import React from 'react';
import {
  IconButton,
  Link,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Tooltip,
  Chip,
} from '@mui/material';
import { makeStyles } from '@mui/styles';
import DeleteIcon from '@mui/icons-material/Delete';
import { IProject, ScanState } from '@api/types';
import { Trans, useTranslation } from 'react-i18next';
import PreviewIcon from '@mui/icons-material/Preview';
import DownloadIcon from '@mui/icons-material/Download';
import SystemUpdateAltOutlinedIcon from '@mui/icons-material/SystemUpdateAltOutlined';

const filter = (items, query) => {
  if (!items) return null;

  const result = items.filter((item) => {
    const name = item.name.toLowerCase();
    return name.includes(query.toLowerCase());
  });

  return result;
};

const format = (date) => {
  return new Date(date).toLocaleDateString('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  });
};

const useStyles = makeStyles((theme) => ({
  md: {
    maxWidth: 130,
    textAlign: 'center',
  },
}));

const isProjectFinished = (project: IProject): boolean => project.scannerState === ScanState.FINISHED;

interface ProjectListProps {
  projects: IProject[];
  searchQuery: string;
  onProjectClick: (project: IProject) => void;
  onProjectDelete: (project: IProject) => void;
  onProjectShowFiles: (project: IProject) => void;
  onProjectDownload: (project: IProject) => void;
  onProjectCreate: () => void;
}

const ProjectList = (props: ProjectListProps) => {
  const classes = useStyles();
  const { t } = useTranslation();

  const { projects, searchQuery } = props;
  const filterProjects = filter(projects, searchQuery);

  return (
    <>
      {projects && projects.length > 0 ? (
        <TableContainer component={Paper}>
          <Table className="projects-table" aria-label="projects table">
            <TableHead>
              <TableRow>
                <TableCell width="50%">{t('Table:Header:Name')}</TableCell>
                <TableCell>{t('Table:Header:Date')}</TableCell>
                <TableCell>{t('Table:Header:TotalFiles')}</TableCell>
                <TableCell width={30} />
              </TableRow>
            </TableHead>
            <TableBody>
              {filterProjects.length !== 0 ? (
                filterProjects.map((project) => (
                  <TableRow
                    className={`
                      ${isProjectFinished(project) ? 'scanning-complete' : 'scanning-not-complete'}
                    `}
                    hover
                    key={project.name}
                    onClick={() => isProjectFinished(project) && props.onProjectClick(project)}
                  >
                    <TableCell component="th" scope="row">
                      <div className="project-name">
                        <span>{project.name}</span>
                      </div>
                    </TableCell>
                    <TableCell>{format(project.date)}</TableCell>
                    <TableCell>{project.files}</TableCell>
                    <TableCell className="row-actions">
                      <div className="btn-actions">
                        <>
                            <Tooltip title={t('Tooltip:ShowFiles')}>
                              <IconButton
                                aria-label="show-files"
                                className="btn-show"
                                disabled={!isProjectFinished(project)}
                                onClick={(event) => {
                                  event.stopPropagation();
                                  props.onProjectShowFiles(project);
                                }}
                                size="large"
                              >
                                <PreviewIcon fontSize="small" />
                              </IconButton>
                            </Tooltip>

                            <Tooltip title={t('Tooltip:CreateFossityPackage')}>
                              <IconButton
                                aria-label="download"
                                disabled={!isProjectFinished(project)}
                                className="btn-download"
                                onClick={(event) => {
                                  event.stopPropagation();
                                  props.onProjectDownload(project);
                                }}
                                size="large"
                              >
                                <SystemUpdateAltOutlinedIcon fontSize="small" />
                              </IconButton>
                            </Tooltip>

                            <Tooltip title={t('Tooltip:RemoveProject')}>
                            <IconButton
                              aria-label="delete"
                              className="btn-delete"
                              onClick={(event) => {
                                event.stopPropagation();
                                props.onProjectDelete(project);
                              }}
                              size="large"
                            >
                              <DeleteIcon fontSize="small" />
                            </IconButton>
                          </Tooltip>
                          </>
                      </div>
                    </TableCell>
                  </TableRow>
                ))
              ) : (
                <TableRow>
                  <TableCell colSpan={4}>
                    <p className="text-center">
                      <Trans
                        i18nKey="Common:NoProjectsFoundWith"
                        components={{
                          strong: <strong/>,
                        }}
                        values={{ searchQuery }}
                      />
                    </p>
                  </TableCell>
                </TableRow>
              )}
            </TableBody>
          </Table>
        </TableContainer>
      ) : !projects ? (
        <p>{t('Common:LoadingProjects')}</p>
      ) : (
        <div className="empty-container">
          <div className="empty-list">
            <h3>{t('Common:NoProjectsFound')}</h3>
            <p>
              <Trans
                i18nKey="Common:StartNewProject"
                components={{
                  link1: <Link onClick={() => props.onProjectCreate()} underline="hover" />,
                }}
                />
            </p>
          </div>
        </div>
      )}
    </>
  );
};

export default ProjectList;
