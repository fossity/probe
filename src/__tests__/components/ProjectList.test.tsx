import React from 'react';
import { configure } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ProjectList from '../../renderer/features/workspace/pages/Components/ProjectList';
import { IProject } from '@api/types';

const props = {
  projects: [],
  searchQuery: '',
  onProjectClick: jest.fn(),
  onProjectDelete: jest.fn(),
  onProjectShowFiles: jest.fn(),
  onProjectCreate: jest.fn(),
  onProjectDownload: jest.fn(),
};

describe('ProjectList', () => {
  it('should render an empty container', () => {
    const component = renderer.create(<ProjectList {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
