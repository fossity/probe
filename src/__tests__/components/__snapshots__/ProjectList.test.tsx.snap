// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList should render an empty container 1`] = `
<div
  className="empty-container"
>
  <div
    className="empty-list"
  >
    <h3>
      Common:NoProjectsFound
    </h3>
    <p />
  </div>
</div>
`;
